### ISSUE TITLE
NetArchTest ignores a nameof expression

### ISSUE THREAD

**Issue by beylkhanovdamir** [BenMorris/NetArchTest#81]
I faced with the strange behavior by NetArchTest when I worked with the `nameof` expression.

I have a simple an architecture test which is below:
`public void DomainLayerDoesNotHaveDependencyOnInfrastructureLayer() {
			TestResult result = Types.InAssembly(typeof(Identity).Assembly)
				.That()
				.ResideInNamespace("StronglyTypedIdentity.Domain")
				.ShouldNot()
				.HaveDependencyOn("StronglyTypedIdentity.Infrastructure")
				.GetResult();

			Assert.That(result.IsSuccessful);
}`
As you see, I want to avoid dependencies on the Infrastructure in the Domain layer (_according to the Onion architecture it's wrong_).

Now I'm going to apply this arch. test for my [pet project ](https://github.com/beylkhanovdamir/StronglyTypedIdentity).

To reproduce a strange behavior with `nameof` I deliberately decided to instantiate an object with `CustomDbContext` type(Infrastructure) in the `CustomType` (Domain). This means that I added a dependency on the **Infrastructure** layer in the Domain layer. 
And now is an interesting moment, cuz, when I'm wrapping `CustomDbContext` into the `nameof` expr., then my arch test is successfully passing, but in fact, it should not. When I don't use nameof, the arch. test result is failed, and that is correct.

Watch my screencast [here](https://www.loom.com/share/3a8938c03f134d14a1d09ee05442711a) to realize an issue

**Comment by NeVeSpl** [comment]
It is happening because `nameof` is replaced with an ordinal string literal during compilation. And NetArchTest is looking for dependencies in compiled code, not source code.

Example: 
https://sharplab.io/#v2:CYLg1APgAgTAjAWAFBQMwAJboMLoN7LpGYZQAs6AsgBQCU6hxBSxr6AbgIYBO6AxugC86AHacAtgFMA9gDNqAZQAu3AJYiA5rQDcjIgF9k+oA===

**Comment by beylkhanovdamir** [comment]
> It is happening because nameof expression is replaced with an ordinal string literal during compilation.
yes, it's. But it's still incorrect I think, isn't it? So would be nice to fix it

**Comment by beylkhanovdamir** [comment]
Let's imagine, that _Domain_, _Infrastructure_, and _API_ are separate assemblies. In this case, I deliberately should add a reference to _Infrastructure_ assembly into the _Domain_ assembly to use `CustomDbContext` in the `CustomType`. It means that NetArchTest will ignore also the reference(dependency) on denied assembly. 
I'm sure that should be fixed...
